WalletWasabi/Wallets/BlockDownloadStats.cs

```diff
@@ -0,0 +1,153 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using NBitcoin.Protocol;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Wallets;
+
+public class BlockDownloadStats
+{
+	private double AverageLastSample { get; set; }
+	private double StandardDeviationLastSample { get; set; }
+	private int ConsecutiveFails { get; set; }
+	private IEnumerable<BlockDl> Sample { get; set; } = Enumerable.Empty<BlockDl>();
+	private List<BlockDl> History { get; } = new();
+	private object Lock { get; } = new();
+	
+	public void AddBlockDl(IPAddress from, double msElapsed, bool success)
+	{
+		lock (Lock)
+		{
+			History.Add((new BlockDl(from, msElapsed, success)));
+			Sample = History.Skip(Math.Max(0, History.Count - Const.SampleSize));
+			(AverageLastSample, StandardDeviationLastSample) = 
+				GetAverageStandardDerivation(Sample.Where(x => x.Success));
+			ConsecutiveFails = success ? 0 : ConsecutiveFails + 1;
+		}
+	}
+
+	private static (double, double) GetAverageStandardDerivation(IEnumerable<BlockDl> collection)
+	{
+		return MathUtils.AverageStandardDeviation(collection.Select(x => x.MsElapsed));
+	}
+	
+	public int NodeTimeoutStrategy(int lastTimeout)
+	{
+		lock (Lock)
+		{
+			int baseTimeout;
+			if (AverageLastSample > 0 && Sample.Count(x => x.Success) > Const.SampleMinSuccesses)
```

Comment:
Can you add more comments on what is happening here? I mean overall to the whole function from line to line. If you can you can extract local variables with meaningful names as well.